themes/default/content/blog/next-level-iac-pulumi-runtime-logic/index.md

```diff
@@ -0,0 +1,428 @@
+
+In this post, the first of a series, we'll show you a few things you can do to take advantage of this reality.
+
+## Run pre-deployment tasks with Pulumi
+
+Since a Pulumi program is just a regular program, you can take actions programmatically before runs --- specifically before Pulumi begins [registering resource declarations](https://www.pulumi.com/docs/concepts/how-pulumi-works).
+
+Suppose you were writing a program to deploy a static website. With a traditional IaC tool, you might model the deployment in steps:
+
+1. Build the website
+1. Deploy the supporting infrastructure (cloud storage, content-delivery front-end)
+1. Copy the content of the website into the infrastructure
+
+To execute these steps, you might reach for something like Bash, writing a shell script similar to the following, which uses [Hugo](https://gohugo.io/), a popular static-site generator, to build the website, Pulumi to provision the infrastructure, and the AWS CLI to upload the content:
+
+```bash
+#!/bin/bash
+
+# Build the website.
+hugo --destination ./public
+
+# Provision the infrastructure.
+pulumi up --yes
+
+# Copy the content of the website into the infrastructure.
+aws s3 sync ./public s3://$(pulumi stack output bucketName)
+```
+
+And this would work --- but it's also unnecessary. You can get rid of the Bash wrapper entirely by just using your chosen language's built-in support for spawning shell processes:
+
+{{< chooser language "typescript,python" />}}
+
+{{% choosable language typescript %}}
+
+```typescript
+import * as aws from "@pulumi/aws";
+import * as childProcess from "child_process";
+
+// Build the website.
+const proc = childProcess.execSync("hugo", { stdio: "inherit", cwd: "./www" });
+
+// Provision a storage bucket for the website.
+const bucket = new aws.s3.Bucket("bucket", {
+    website: {
+        indexDocument: "index.html"
+    },
+});
+
+// Copy the website home page into the bucket.
+const homepage = new aws.s3.BucketObject("index.html", {
+    bucket: bucket.id,
+    content: fs.readFileSync("./www/public/index.html", "utf-8"),
+    contentType: "text/html",
+    acl: "public-read",
+});
+```
+
+{{% /choosable %}}
+
+{{% choosable language python %}}
+
+```python
+import pulumi_aws as aws
+import subprocess
+
+# Build the website.
+result = subprocess.run(["hugo"], stdout=subprocess.PIPE, cwd="./www")
+
+# Provision a storage bucket for the website.
+bucket = aws.s3.Bucket(
+    "bucket", website=aws.s3.BucketWebsiteArgs(index_document="index.html")
+)
+
+# Copy the website home page into the bucket
+homepage = aws.s3.BucketObject(
+    "index.html",
+    bucket=bucket.id,
+    content=open("./www/public/index.html", "r").read(),
+    content_type="text/html",
+    acl="public-read",
+)
+```
+
+{{% /choosable %}}
+
+Because these processes run synchronously, they finish before the deployment engine gets going, terminating the containing process on error and logging any relevant output to the console:
```

Comment:
Can you make this more clear. You speak about it being an error, but it seems like a successful run